Bitwise builtin (#69)
lambdaclass/cairo-vm.go: README.md

[Next sections: Implementing each builtin runner]

### Implementing each builtin runner 

#### Bitwise builtin

This builtin provides a way to work with the basic bit operations `and`, `or` and `xor`. It implements the basic builtin interface methods:

> 

Reviewer: Add BitWiseBuiltinRunner type definton

```go
func (b *BitwiseBuiltinRunner) Base() memory.Relocatable {
	return r.base
}

func (b *BitwiseBuiltinRunner) Name() string {
	return "range_check"
}
```

For the `InitializeSegments` we just add a segment to the memory and store in the base attribute, the first adress of the segment. 

```go
func (b *BitwiseBuiltinRunner) InitializeSegments(segments *memory.MemorySegmentManager) {
	r.base = segments.AddSegment()
}
```

we also have `InitialStack` method that returns a stack the base address appended

```go 
func (b *BitwiseBuiltinRunner) InitialStack() []memory.MaybeRelocatable {
	if b.included {
		return []memory.MaybeRelocatable{*memory.NewMaybeRelocatableRelocatable(b.base)}
	} else {
		return []memory.MaybeRelocatable{}
	}
}
```

> The method `DeducedMemoryCell` fetches from memory the operands and  performes the operations. 

Reviewer: ```suggestion
The method `DeducedMemoryCell` fetches the operands from memory and performs the following operations.:
```

> - If the index is lesser than two then the methods just return nil

Reviewer: ```suggestion
- If the index is less than the number of input cells then the methods just return nil
```

> - After the operands are fetch it is checked that both are felts because bitwise can not be perform on relocatables or nil values.

Reviewer: ```suggestion
- After the operands are fetched it checks that both of them are felts because bitwise can not be performed on relocatable or nil values.
```

- If the number of bits of any operands is greater than TOTAL_N_BITS the method fails because we are out of the field.
- The index is used to know which bitwise operation is going to be performed, if is 2 then `and` is executed, if is 3 then `xor` and if is a 4 then `or` is executed
- Otherwise nil value is returned

> ```go 

Reviewer: Add the definition of the constants to this code block

[Next sections: Implementing each builtin runner]

> ### Implementing each builtin runner 

Reviewer: ```suggestion
#### Implementing each builtin runner 
```
This has the same hierarchy as Write your own Vm title

> #### Bitwise builtin

Reviewer: ```suggestion
##### Bitwise builtin
```

This builtin provides a way to work with the basic bit operations `and`, `or` and `xor`. It implements the basic builtin interface methods:

```go
type BitwiseBuiltinRunner struct {
	base     memory.Relocatable
	included bool
}
```
the getter methods: 

```go
func (b *BitwiseBuiltinRunner) Base() memory.Relocatable {
	return r.base
}

func (b *BitwiseBuiltinRunner) Name() string {
	return "range_check"
}
```

> For the `InitializeSegments` we just add a segment to the memory and store in the base attribute, the first adress of the segment. 

Reviewer 2: ```suggestion
For the `InitializeSegments` we just add a segment to the memory and store in the base attribute, which is the first address of the segment. 
```

Reviewer 3: I think it would be nicer something like: 

```suggestion
In the `InitializeSegments` method, we add a new segment to the memory and set the base attribute as the first segment's address. 
```

Reviewer 2: Agree

```go
func (b *BitwiseBuiltinRunner) InitializeSegments(segments *memory.MemorySegmentManager) {
	r.base = segments.AddSegment()
}
```

we also have `InitialStack` method that returns a stack the base address appended

```go 
func (b *BitwiseBuiltinRunner) InitialStack() []memory.MaybeRelocatable {
	if b.included {
		return []memory.MaybeRelocatable{*memory.NewMaybeRelocatableRelocatable(b.base)}
	} else {
		return []memory.MaybeRelocatable{}
	}
}
```

> The method `DeducedMemoryCell` fetches the operands from memory performes the following operations: 

Reviewer 2: ```suggestion
The method `DeducedMemoryCell` fetches the operands from memory and performs the following operations: 
```

> - If the index is less than the number of input cells then the methods just return nil

Reviewer 2: ```suggestion
- If the index is less than the number of input cells then the method returns nil
```

- After the operands are fetched it is checks that both of them are felts because bitwise can not be performed on relocatable or nil values.

> - If the number of bits of any operands is greater than TOTAL_N_BITS the method fails because we are out of the field.

Reviewer 2: ```suggestion
- If the number of bits of any operand is greater than `TOTAL_N_BITS` the method fails because we are out of the field.
```

func (b *BitwiseBuiltinRunner) InitializeSegments(segments *memory.MemorySegmentManager) {
	r.base = segments.AddSegment()
}
```

we also have `InitialStack` method that returns a stack the base address appended

```go 
func (b *BitwiseBuiltinRunner) InitialStack() []memory.MaybeRelocatable {
	if b.included {
		return []memory.MaybeRelocatable{*memory.NewMaybeRelocatableRelocatable(b.base)}
	} else {
		return []memory.MaybeRelocatable{}
	}
}
```

The method `DeducedMemoryCell` fetches the operands from memory and performs the following operations: 
- If the index is less than the number of input cells then the method returns nil
- After the operands are fetched it is checks that both of them are felts because bitwise can not be performed on relocatable or nil values.
- If the number of bits of any operand is greater than `TOTAL_N_BITS` the method fails because we are out of the field.
- The index is used to know which bitwise operation is going to be performed, if is 2 then `and` is executed, if is 3 then `xor` and if is a 4 then `or` is executed
- Otherwise nil value is returned

```go 
const CELLS_PER_INSTANCE = 5
const TOTAL_N_BITS = 251
const INPUT_CELLS_PER_INSTANCE = 2

> func (b *BitwiseBuiltinRunner) DeduceMemoryCell(address memory.Relocatable, segments *memory.Memory) (*memory.MaybeRelocatable, error) {

Reviewer: Please update this

#### Hints

#### Implementing each builtin runner 

> 

Reviewer: Remove
